File: src/Coda.Agent/Tasks/OutputRing.cs
using System.Text;

namespace Coda.Agent.Tasks;

/// <summary>
/// Bounded, thread-safe, drop-oldest text buffer. Tracks absolute character
/// offsets so incremental readers can detect when data they had not yet read
/// was evicted (Truncated == true).
///
/// Memory bounding is enforced two ways:
/// <list type="bullet">
/// <item>Tiny adjacent appends are coalesced into bounded-size chunks (see
/// <see cref="_chunkTarget"/>) so a stream of 1-byte appends produces a small,
/// bounded number of nodes rather than one node per append.</item>
/// <item>Whole chunks are evicted from the front once the total exceeds the cap.
/// A single append larger than the cap is trimmed to the newest UTF-8-valid
/// suffix that fits, so no unbounded payload is ever retained.</item>
/// </list>
///
/// Absolute UTF-16 cursor semantics are preserved: <see cref="TotalChars"/>
/// always counts every character ever appended (including trimmed/evicted ones),
/// and <see cref="DroppedChars"/> counts characters removed from the front. All
/// chunk boundaries (append, eviction, and oversized trimming) fall on Unicode
/// code-point boundaries, so buffered/returned strings are always valid UTF-16.
/// </summary>
internal sealed class OutputRing
{
    public const long DefaultMaxBytes = 1L << 20; // 1 MiB

    /// <summary>Upper bound on the payload size we coalesce into a single chunk.</summary>
    private const int MaxChunkTarget = 16 * 1024; // 16 KiB

    private sealed class Chunk
    {
        public Chunk(string text, int byteLength, long startChar)
        {
            Text = new StringBuilder(text);
            ByteLength = byteLength;
            StartChar = startChar;
        }

        public StringBuilder Text { get; }
        public int ByteLength { get; set; }
        public long StartChar { get; set; }
        public int CharLength => Text.Length;
    }

    private readonly object _gate = new();
    private readonly long _maxBytes;
    private readonly int _chunkTarget;
    private readonly LinkedList<Chunk> _chunks = new();
    private long _byteLength;
    private long _totalChars;   // absolute chars ever appended
    private long _droppedChars; // absolute chars evicted from the front

    public OutputRing(long maxBytes = DefaultMaxBytes)
    {
        if (maxBytes <= 0) throw new ArgumentOutOfRangeException(nameof(maxBytes));
        _maxBytes = maxBytes;
        // Keep chunk granularity well below the cap so eviction stays reasonably
        // fine-grained, while capping tiny-append coalescing at 16 KiB payloads.
        _chunkTarget = (int)Math.Clamp(maxBytes / 16, 1, MaxChunkTarget);
    }

    /// <summary>Total characters ever appended, including evicted ones.</summary>
    public long TotalChars { get { lock (_gate) { return _totalChars; } } }

    /// <summary>Characters evicted from the front so far.</summary>
    public long DroppedChars { get { lock (_gate) { return _droppedChars; } } }

    /// <summary>Number of internal chunk nodes. Exposed for memory-bounding tests.</summary>
    internal int ChunkCount { get { lock (_gate) { return _chunks.Count; } } }

    public void Append(string text)
    {
        if (string.IsNullOrEmpty(text)) return;
        var bytes = Encoding.UTF8.GetByteCount(text);
        lock (_gate)
        {
            var startChar = _totalChars;
            _totalChars += text.Length;

            var last = _chunks.Last?.Value;
            if (last is not null && last.ByteLength < _chunkTarget)
            {
                // Coalesce into the still-growing tail chunk (amortized O(1)).
                last.Text.Append(text);
                last.ByteLength += bytes;
            }
            else
            {
                _chunks.AddLast(new Chunk(text, bytes, startChar));
            }
            _byteLength += bytes;

            EvictWhileOverCap();
            TrimNewestIfOversized();
        }
    }

    private void EvictWhileOverCap()
    {
        // Keep at least the most recent chunk even if it alone exceeds the cap;
        // an oversized sole chunk is handled by TrimNewestIfOversized instead.
        while (_byteLength > _maxBytes && _chunks.Count > 1)
        {
            var first = _chunks.First!.Value;
            _chunks.RemoveFirst();
            _byteLength -= first.ByteLength;
            _droppedChars = first.StartChar + first.CharLength;
        }
    }

    /// <summary>
    /// When the only remaining chunk still exceeds the cap (an oversized single
    /// append), trim its prefix so only the newest UTF-8-valid suffix that fits
    /// <see cref="_maxBytes"/> is retained. TotalChars is unchanged (it counts
    /// the whole input); DroppedChars advances by the trimmed prefix length.
    /// </summary>
    private void TrimNewestIfOversized()
    {
        if (_chunks.Count != 1) return;
        var only = _chunks.First!.Value;
        if (only.ByteLength <= _maxBytes) return;

        var start = SuffixStartFitting(only.Text, _maxBytes);
        if (start <= 0) return;

        var retainedBytes = Utf8Bytes(only.Text, start, only.Text.Length - start);
        only.Text.Remove(0, start);
        _byteLength -= only.ByteLength - retainedBytes;
        only.ByteLength = retainedBytes;
        only.StartChar += start;
        _droppedChars = only.StartChar;
    }

    /// <summary>
    /// Returns the char index of the newest suffix of <paramref name="s"/> whose
    /// UTF-8 encoding fits in <paramref name="maxBytes"/>, cut on a code-point
    /// boundary (never splitting a surrogate pair). If even the last code point
    /// alone exceeds the cap, that whole rune is still retained (matching Peek's
    /// whole-rune rule) rather than returning an empty, lossy result.
    /// </summary>
    private static int SuffixStartFitting(StringBuilder s, long maxBytes)
    {
        long bytes = 0;
        var i = s.Length;
        while (i > 0)
        {
            var step = 1;
            if (char.IsLowSurrogate(s[i - 1]) && i >= 2 && char.IsHighSurrogate(s[i - 2]))
            {
                step = 2;
            }

            var cpBytes = Utf8Bytes(s, i - step, step);
            if (bytes + cpBytes > maxBytes)
            {
                if (i == s.Length) i -= step; // guarantee at least the newest rune
                break;
            }

            bytes += cpBytes;
            i -= step;
        }
        return i;
    }

    private static int Utf8Bytes(StringBuilder s, int start, int length) =>
        Encoding.UTF8.GetByteCount(s.ToString(start, length));

    /// <summary>
    /// Reads all buffered text at or after the absolute character offset
    /// <paramref name="cursor"/>. Returns the concatenated text, the next cursor
    /// to pass on the following call, and whether earlier data was evicted before
    /// the caller could read it (cursor &lt; DroppedChars). Negative cursors are
    /// clamped to zero.
    /// </summary>
    public (string Text, long NextCursor, bool Truncated) ReadFrom(long cursor)
    {
        lock (_gate)
        {
            return ReadFromNoLock(cursor);
        }
    }

    /// <summary>Returns the last <paramref name="maxChars"/> characters currently buffered.</summary>
    public string Peek(int maxChars)
    {
        if (maxChars <= 0) return string.Empty;
        lock (_gate)
        {
            var start = Math.Max(_droppedChars, _totalChars - maxChars);
            var (text, _, _) = ReadFromNoLock(start);

            // The requested boundary may land on the low surrogate of a pair.
            // Widen to include the preceding high surrogate so we return the whole
            // rune (2 code units even for maxChars == 1) instead of malformed
            // UTF-16. If the high half was already evicted, drop the lone low
            // surrogate rather than emit it.
            if (text.Length > 0 && char.IsLowSurrogate(text[0]))
            {
                if (start - 1 >= _droppedChars)
                {
                    (text, _, _) = ReadFromNoLock(start - 1);
                }
                else
                {
                    text = text[1..];
                }
            }

            return text;
        }
    }

    private (string Text, long NextCursor, bool Truncated) ReadFromNoLock(long cursor)
    {
        if (cursor < 0) cursor = 0;
        var truncated = cursor < _droppedChars;
        var from = Math.Max(cursor, _droppedChars);
        if (from >= _totalChars) return (string.Empty, _totalChars, truncated);

        var sb = new StringBuilder();
        foreach (var chunk in _chunks)
        {
            var chunkEnd = chunk.StartChar + chunk.CharLength;
            if (chunkEnd <= from) continue;
            var localStart = (int)Math.Max(0, from - chunk.StartChar);
            sb.Append(chunk.Text.ToString(localStart, chunk.CharLength - localStart));
        }
        return (sb.ToString(), _totalChars, truncated);
    }
}

File: rust/crates/coda-agent/src/tasks/output_ring.rs
//! Bounded, thread-safe, drop-oldest text buffer with absolute char cursors.
//!
//! Adjacent tiny appends are coalesced into bounded-size chunks so a stream of
//! 1-byte appends produces only O(capacity / chunk_target) nodes rather than
//! one node per byte. Chunks are evicted from the front once total byte length
//! exceeds the cap. A single append larger than the cap is trimmed to the newest
//! valid UTF-8 suffix that fits.
//!
//! Absolute UTF-8-char cursor semantics: `total_chars` counts every character
//! ever appended (including trimmed/evicted), and `dropped_chars` counts
//! characters removed from the front. All chunk boundaries fall on Unicode
//! code-point boundaries.

use std::collections::VecDeque;
use std::sync::Mutex;

/// 1 MiB default — matches the C# `DefaultMaxBytes`.
pub const DEFAULT_MAX_BYTES: u64 = 1 << 20;

/// Upper bound on the payload coalesced into a single chunk (16 KiB).
const MAX_CHUNK_TARGET: usize = 16 * 1024;

struct Chunk {
    text: String,
    byte_len: usize,
    /// Absolute char offset of the first character in this chunk.
    start_char: u64,
}

impl Chunk {
    fn char_len(&self) -> u64 {
        self.text.chars().count() as u64
    }
}

struct RingState {
    chunks: VecDeque<Chunk>,
    byte_len: u64,
    /// Total chars ever appended, including dropped/trimmed.
    total_chars: u64,
    /// Chars removed from the front so far.
    dropped_chars: u64,
}

pub struct OutputRing {
    state: Mutex<RingState>,
    max_bytes: u64,
    /// Target chunk payload size for coalescing (≤ `MAX_CHUNK_TARGET`).
    chunk_target: usize,
}

impl OutputRing {
    pub fn new(max_bytes: u64) -> Self {
        assert!(max_bytes > 0, "max_bytes must be positive");
        let chunk_target = ((max_bytes / 16) as usize).clamp(1, MAX_CHUNK_TARGET);
        Self {
            state: Mutex::new(RingState {
                chunks: VecDeque::new(),
                byte_len: 0,
                total_chars: 0,
                dropped_chars: 0,
            }),
            max_bytes,
            chunk_target,
        }
    }

    /// Total characters ever appended (including dropped/trimmed).
    pub fn total_chars(&self) -> u64 {
        self.state.lock().unwrap().total_chars
    }

    /// Characters removed from the front so far.
    pub fn dropped_chars(&self) -> u64 {
        self.state.lock().unwrap().dropped_chars
    }

    /// Number of internal chunk nodes (for bounding tests).
    #[cfg(test)]
    pub fn chunk_count(&self) -> usize {
        self.state.lock().unwrap().chunks.len()
    }

    /// Append text, evicting old chunks as needed to stay within the byte cap.
    pub fn append(&self, text: &str) {
        if text.is_empty() {
            return;
        }
        let bytes = text.len(); // UTF-8 byte count
        let mut s = self.state.lock().unwrap();
        let start_char = s.total_chars;
        s.total_chars += text.chars().count() as u64;

        // Coalesce into the still-growing tail chunk when it is under the target.
        if let Some(last) = s.chunks.back_mut() {
            if last.byte_len < self.chunk_target {
                last.text.push_str(text);
                last.byte_len += bytes;
                s.byte_len += bytes as u64;
                evict_while_over_cap(&mut s, self.max_bytes);
                trim_newest_if_oversized(&mut s, self.max_bytes);
                return;
            }
        }

        s.chunks.push_back(Chunk {
            text: text.to_owned(),
            byte_len: bytes,
            start_char,
        });
        s.byte_len += bytes as u64;
        evict_while_over_cap(&mut s, self.max_bytes);
        trim_newest_if_oversized(&mut s, self.max_bytes);
    }

    /// Read all buffered text at or after the absolute character offset `cursor`.
    ///
    /// Returns `(text, next_cursor, truncated)` where:
    /// - `text` is the concatenated text from `cursor` onward
    /// - `next_cursor` is the cursor to pass on the next call
    /// - `truncated` is true when data before `cursor` was already evicted
    pub fn read_from(&self, cursor: u64) -> (String, u64, bool) {
        let s = self.state.lock().unwrap();
        read_from_locked(&s, cursor)
    }

    /// Returns the last `max_chars` characters currently buffered.
    pub fn peek(&self, max_chars: usize) -> String {
        if max_chars == 0 {
            return String::new();
        }
        let s = self.state.lock().unwrap();
        let start = s.dropped_chars.max(s.total_chars.saturating_sub(max_chars as u64));
        let (text, _, _) = read_from_locked(&s, start);

        // Trim from the front if we ended up with more than max_chars (the start
        // calculation is char-approximate; read_from_locked returns by byte slice).
        let count = text.chars().count();
        if count > max_chars {
            let skip = count - max_chars;
            let byte_offset = text
                .char_indices()
                .nth(skip)
                .map(|(i, _)| i)
                .unwrap_or(text.len());
            text[byte_offset..].to_owned()
        } else {
            text
        }
    }
}

// ── Internal helpers (operate on the locked state) ───────────────────────────

fn read_from_locked(s: &RingState, cursor: u64) -> (String, u64, bool) {
    let truncated = cursor < s.dropped_chars;
    let from = cursor.max(s.dropped_chars);
    if from >= s.total_chars {
        return (String::new(), s.total_chars, truncated);
    }

    let mut result = String::new();
    for chunk in &s.chunks {
        let chunk_end = chunk.start_char + chunk.char_len();
        if chunk_end <= from {
            continue;
        }
        let local_start_chars = (from.saturating_sub(chunk.start_char)) as usize;
        let byte_start = chunk
            .text
            .char_indices()
            .nth(local_start_chars)
            .map(|(i, _)| i)
            .unwrap_or(chunk.text.len());
        result.push_str(&chunk.text[byte_start..]);
    }
    (result, s.total_chars, truncated)
}

/// Evict whole chunks from the front until byte_len ≤ max_bytes.
/// Always keeps at least the last chunk so an oversized sole chunk can be
/// trimmed by `trim_newest_if_oversized` instead.
fn evict_while_over_cap(s: &mut RingState, max_bytes: u64) {
    while s.byte_len > max_bytes && s.chunks.len() > 1 {
        let first = s.chunks.pop_front().unwrap();
        s.byte_len -= first.byte_len as u64;
        s.dropped_chars = first.start_char + first.char_len();
    }
}

/// When the sole remaining chunk still exceeds the cap (single oversized append),
/// trim its prefix to retain only the newest UTF-8-valid suffix that fits.
fn trim_newest_if_oversized(s: &mut RingState, max_bytes: u64) {
    if s.chunks.len() != 1 {
        return;
    }
    let only = s.chunks.back_mut().unwrap();
    if only.byte_len as u64 <= max_bytes {
        return;
    }

    let start_byte = suffix_start_fitting(&only.text, max_bytes as usize);
    if start_byte == 0 {
        return;
    }

    let dropped_chars = only.text[..start_byte].chars().count() as u64;
    let retained_bytes = only.text.len() - start_byte;
    let retained = only.text[start_byte..].to_owned();
    s.byte_len -= (only.byte_len - retained_bytes) as u64;
    only.text = retained;
    only.byte_len = retained_bytes;
    only.start_char += dropped_chars;
    s.dropped_chars = only.start_char;
}

/// Returns the byte index of the newest suffix of `s` whose UTF-8 encoding fits
/// in `max_bytes`, cut on a code-point boundary. Returns 0 if even the last
/// code point alone exceeds the cap (retain the whole string — mirrors C#).
fn suffix_start_fitting(s: &str, max_bytes: usize) -> usize {
    let mut bytes: usize = 0;
    let mut result = s.len(); // default: trim everything (start at end)

    // Walk code points from the back, accumulating byte counts.
    for (byte_idx, ch) in s.char_indices().rev() {
        let cp_bytes = ch.len_utf8();
        if bytes + cp_bytes > max_bytes {
            // This code point would push us over.
            if result == s.len() {
                // The very last rune alone exceeds the cap — retain it anyway.
                result = byte_idx;
            }
            break;
        }
        bytes += cp_bytes;
        result = byte_idx;
    }

    result
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn ring(max_bytes: u64) -> OutputRing {
        OutputRing::new(max_bytes)
    }

    // ── basic ─────────────────────────────────────────────────────────────────

    #[test]
    fn fresh_ring_is_empty() {
        let r = ring(1024);
        assert_eq!(r.total_chars(), 0);
        assert_eq!(r.dropped_chars(), 0);
        let (text, next, trunc) = r.read_from(0);
        assert!(text.is_empty());
        assert_eq!(next, 0);
        assert!(!trunc);
    }

    #[test]
    fn append_increments_total_chars() {
        let r = ring(1024);
        r.append("hello");
        assert_eq!(r.total_chars(), 5);
        r.append(" world");
        assert_eq!(r.total_chars(), 11);
    }

    #[test]
    fn read_from_zero_returns_all_text() {
        let r = ring(1024);
        r.append("hello world");
        let (text, next, trunc) = r.read_from(0);
        assert_eq!(text, "hello world");
        assert_eq!(next, 11);
        assert!(!trunc);
    }

    #[test]
    fn read_from_mid_cursor_returns_tail() {
        let r = ring(1024);
        r.append("abcdef");
        let (text, next, trunc) = r.read_from(3);
        assert_eq!(text, "def");
        assert_eq!(next, 6);
        assert!(!trunc);
    }

    #[test]
    fn read_from_past_end_returns_empty() {
        let r = ring(1024);
        r.append("abc");
        let (text, next, trunc) = r.read_from(100);
        assert!(text.is_empty());
        assert_eq!(next, 3);
        assert!(!trunc);
    }

    #[test]
    fn peek_returns_last_n_chars() {
        let r = ring(4096);
        r.append("abcdefgh");
        assert_eq!(r.peek(3), "fgh");
        assert_eq!(r.peek(8), "abcdefgh");
        assert_eq!(r.peek(100), "abcdefgh");
        assert_eq!(r.peek(0), "");
    }

    // ── memory bounding ───────────────────────────────────────────────────────

    /// Emitting far more bytes than the cap must never grow memory past the cap.
    #[test]
    fn large_volume_stays_within_cap() {
        let cap: u64 = 4096;
        let r = ring(cap);

        // Append 100× the cap in 512-byte chunks.
        let chunk = "x".repeat(512);
        for _ in 0..800 {
            r.append(&chunk);
        }

        let s = r.state.lock().unwrap();
        assert!(
            s.byte_len <= cap,
            "byte_len {} exceeded cap {}",
            s.byte_len,
            cap
        );
        // Overflow must be accounted for.
        assert!(s.dropped_chars > 0, "expected some drops after overflow");
    }

    #[test]
    fn chunk_count_stays_bounded() {
        let cap: u64 = 4096;
        let r = ring(cap);
        for _ in 0..10_000 {
            r.append("x");
        }
        // At most a few chunks should survive.
        assert!(r.chunk_count() < 32, "too many chunks: {}", r.chunk_count());
    }

    // ── truncation flag ────────────────────────────────────────────────────────

    #[test]
    fn truncated_is_true_when_cursor_is_before_dropped() {
        let cap: u64 = 100;
        let r = ring(cap);
        r.append(&"x".repeat(200));
        let dropped = r.dropped_chars();
        assert!(dropped > 0);
        let (_, _, trunc) = r.read_from(dropped - 1);
        assert!(trunc);
        let (_, _, trunc) = r.read_from(dropped);
        assert!(!trunc);
    }

    // ── consecutive reads are cursor-monotonic ────────────────────────────────

    #[test]
    fn incremental_reads_cover_all_output() {
        let r = ring(4096);
        r.append("aaa");
        let (t1, c1, _) = r.read_from(0);
        assert_eq!(t1, "aaa");
        r.append("bbb");
        let (t2, c2, _) = r.read_from(c1);
        assert_eq!(t2, "bbb");
        r.append("ccc");
        let (t3, _, _) = r.read_from(c2);
        assert_eq!(t3, "ccc");
    }

    // ── single oversized append is trimmed, not rejected ─────────────────────

    #[test]
    fn oversized_single_append_is_trimmed_to_cap() {
        let cap: u64 = 64;
        let r = ring(cap);
        r.append(&"z".repeat(1000));
        let s = r.state.lock().unwrap();
        assert!(s.byte_len <= cap, "byte_len {} > cap {}", s.byte_len, cap);
    }
}
